### PR TITLE
feat: Add catalog base page for customizations

### PR DESCRIPTION
**Pull request by hostsamurai**
## Hey, I just made a Pull Request!

Following up on some of the [advice](https://github.com/backstage/backstage/pull/4845#discussion_r597591980) from #4845, this PR offers a way of letting
users customize the catalog pages by offering a "base" catalog page with the
option of passing in a custom table component with whatever columns that users
may want to render. The new base catalog page component is meant to be used in
other plugins, such as `api-docs`, as a way to avoid having separate
implementations of the catalog page and table components. #4845 includes a
proof-of-concept implementation for the `api-docs` plugin that replaces its
catalog page and table with ones from this PR. An example of a custom catalog
page and table that can be rendered as a result of this work looks like the
following:

113224971-2f17e200-9252-11eb-81ed-a648894f86a5

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/catalog/src/components/CatalogPage/CatalogPage.tsx [hunk 1]
```diff
@@ -14,213 +14,49 @@
  * limitations under the License.
  */
 
-import {
```
Comment (hostsamurai) [inline plugins/catalog/src/components/CatalogPage/CatalogPage.tsx:17]: Much of the content from this component was moved to `CatalogPageBase`.

File: plugins/catalog/src/components/useOwnUser.ts [hunk 2]
```diff
@@ -30,12 +30,12 @@ export function useOwnUser(): AsyncState<UserEntity | undefined> {
   // TODO: get the full entity (or at least the full entity name) from the
   // identityApi
   return useAsync(
-    () =>
-      catalogApi.getEntityByName({
+    async () =>
+      await catalogApi.getEntityByName({
```
Comment (hostsamurai) [inline plugins/catalog/src/components/useOwnUser.ts:34]: Had to use `async/await` here, otherwise the hook would blow up the catalog page.

**Comment by kuangp** [comment]
Does this overlap with efforts in #5643?

**Comment by timbonicus** [comment]
Yep, I think this competes a bit with #5643 which takes a more composable approach. Great minds think alike though, that PR is also designed to be applicable to the api-docs page, scaffolder page, and other places where entities are fetched from the catalog-backend.

Potentially also overlapping with #5593 that was merged allowing custom columns for the catalog table.

**Comment by hostsamurai** [comment]
I see. I was unaware of #5643 prior to publishing this PR. I originally set out to make the catalog page in `api-docs` customizable, but that grew into a larger effort based on the conversation in #4845. At the time, consensus seemed to favor having a base catalog page that would allow reuse across the different catalog pages. I'm curious to see if this kind of thinking changed. I did see #5593, however, and the changes included here build on top of it.

**Comment by hostsamurai** [comment]
Closing this as #5643 and #5984 provide the functionality that this PR was paving the way to implement.

**Pull request closed without merging**